mypy/messages.py

```diff
@@ -1396,27 +1385,109 @@ def format_type_bare(typ: Type, verbosity: int = 0) -> str:
         return 'object'
 
 
+def collect_all_instances(t: Type) -> List[Instance]:
+    """Return all instances that `t` contains (including `t`)
+
+    This is similar to collect_all_inner_types from typeanal but only
+    returns instances and will recurse into fallbacks when they affect
+    how a type is formatted.
+    """
+    return CollectAllInstancesQuery().query_types([t])
+
+
+class CollectAllInstancesQuery(TypeQuery[List[Instance]]):
+    def __init__(self) -> None:
+        super().__init__(self.combine_lists_strategy)
+
+    def visit_literal_type(self, t: LiteralType) -> List[Instance]:
+        return self.query_types([t.fallback])
+
+    def visit_tuple_type(self, t: TupleType) -> List[Instance]:
+        return self.query_types(t.items + [t.partial_fallback])
+
+    def visit_typeddict_type(self, t: TypedDictType) -> List[Instance]:
+        return self.query_types(list(t.items.values()) + [t.fallback])
+
+    def query_types(self, types: Iterable[Type]) -> List[Instance]:
```

Comment:
Overriding `query_types` makes this hard to reason about. It took me a while to figure out why this works at all. This also seems to remove the cyclic type avoidance in the base class. I'd suggest not overriding this (as it's not really designed to be overridden) but to add additional visit methods for `Instance` and `Callable` at least (there may be others). You could also simplify `collect_all_instances` to just use `accept`, which is the normal entry point, instead of calling `query_types`, which is meant to be an internal helper.

Reply:
I'm going to rework it to use `TypeTraverserVisitor`